src/edu/washington/escience/myria/parallel/Server.java

```diff
@@ -913,8 +919,103 @@ public DatasetStatus ingestDataset(
   }
 
   /**
-   * @param relationKey the relationalKey of the dataset to add
-   * @param schema the schema of the dataset to add
+   * Parallel Ingest
+   *
+   * @param relationKey the name of the dataset.
+   * @param workersToIngest restrict the workers to ingest data (null for all)
+   * @throws URIException
+   * @throws DbException
+   * @throws InterruptedException
+   */
+  public DatasetStatus parallelIngestDataset(
+      final RelationKey relationKey,
+      final Schema schema,
+      @Nullable final Character delimiter,
+      @Nullable final Character quote,
+      @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines,
+      final String S3URI,
+      final Set<Integer> workersToIngest)
+      throws URIException, DbException, InterruptedException {
+    /* Figure out the workers we will use */
+
+    Set<Integer> actualWorkers = workersToIngest;
+    AmazonS3Source s3Source = new AmazonS3Source(S3URI);
+    long fileSize = s3Source.getFileSize();
+
+    /* Determine the number of workers to ingest and the partition size */
+    long partitionSize = 0;
+    int[] workersArray;
+
+    if (workersToIngest == null) {
+      int[] allWorkers = MyriaUtils.integerSetToIntArray(getAliveWorkers());
+      int totalNumberOfWorkersToIngest = 0;
+      for (int i = allWorkers.length; i >= 1; i--) {
+        totalNumberOfWorkersToIngest = i;
+        long currentPartitionSize = fileSize / i;
+        if (currentPartitionSize > MyriaConstants.WORKER_PARALLEL_INGEST_MINIMUM_PARTITION_SIZE
+            || totalNumberOfWorkersToIngest == 1) {
+          partitionSize = currentPartitionSize;
+          break;
+        }
+      }
+      workersArray = Arrays.copyOfRange(allWorkers, 0, totalNumberOfWorkersToIngest);
+    } else {
+      Preconditions.checkArgument(actualWorkers.size() > 0, "Must use > 0 workers");
+      workersArray = MyriaUtils.integerSetToIntArray(actualWorkers);
+      partitionSize = fileSize / workersArray.length;
+    }
+
+    Map<Integer, SubQueryPlan> workerPlans = new HashMap<>();
+    int workerCounterID = 1;
+    for (int workerID : workersArray) {
```

Comment:
It would be simpler just to use an old-fashioned `for` loop over the array indices.